pkg/infoschema/infoschema_v2.go

```diff
@@ -276,6 +289,116 @@ func (isd *Data) deleteDB(dbInfo *model.DBInfo, schemaVersion int64) {
+	}
+	if h.prev != nil &&
+		h.prev.referredFKInfo.ChildSchema.L == item.referredFKInfo.ChildSchema.L &&
+		h.prev.referredFKInfo.ChildTable.L == item.referredFKInfo.ChildTable.L &&
+		h.prev.referredFKInfo.ChildFKName.L == item.referredFKInfo.ChildFKName.L {
+		return true
+	}
+	h.prev = item
+	if !item.tomb {
+		h.referredFKInfos = append(h.referredFKInfos, item.referredFKInfo)
+	}
+	return true
+}
+
+func (isd *Data) getTableReferredForeignKeys(schema, table string, schemaMetaVersion int64) []*model.ReferredFKInfo {
+	helper := referredForeignKeysHelper{
+		start:           referredForeignKeyItem{dbName: schema, tableName: table, schemaVersion: math.MaxInt64},
+		schemaVersion:   schemaMetaVersion,
+		referredFKInfos: make([]*model.ReferredFKInfo, 0),
+	}
+	isd.referredForeignKeys.Load().DescendLessOrEqual(&helper.start, helper.onItem)
+
+	sort.Slice(helper.referredFKInfos, func(i, j int) bool {
+		if helper.referredFKInfos[i].ChildSchema.L != helper.referredFKInfos[j].ChildSchema.L {
+			return helper.referredFKInfos[i].ChildSchema.L < helper.referredFKInfos[j].ChildSchema.L
+		}
+		if helper.referredFKInfos[i].ChildTable.L != helper.referredFKInfos[j].ChildTable.L {
+			return helper.referredFKInfos[i].ChildTable.L < helper.referredFKInfos[j].ChildTable.L
+		}
+		return helper.referredFKInfos[i].ChildFKName.L < helper.referredFKInfos[j].ChildFKName.L
+	})
+	return helper.referredFKInfos
+}
+
+func (isd *Data) addReferredForeignKeys(schema ast.CIStr, tbInfo *model.TableInfo, schemaMetaVersion int64) {
+	for _, fk := range tbInfo.ForeignKeys {
+		if fk.Version < model.FKVersion1 {
+			continue
+		}
+		// found if already exists
+		referredFKInfos := isd.getTableReferredForeignKeys(fk.RefSchema.L, fk.RefTable.L, schemaMetaVersion)
+		found := false
+		for _, referredFK := range referredFKInfos {
+			if referredFK.ChildSchema.L == schema.L && referredFK.ChildTable.L == tbInfo.Name.L && referredFK.ChildFKName.L == fk.Name.L {
+				found = true
+				break
+			}
+		}
+		if found {
+			continue
+		}
+		btreeSet(&isd.referredForeignKeys, &referredForeignKeyItem{
+			dbName:        fk.RefSchema.L,
+			tableName:     fk.RefTable.L,
+			schemaVersion: schemaMetaVersion,
+			referredFKInfo: &model.ReferredFKInfo{
+				Cols:        fk.RefCols,
+				ChildSchema: schema,
+				ChildTable:  tbInfo.Name,
+				ChildFKName: fk.Name,
+			},
+		})
+	}
+}
+
+func (isd *Data) deleteReferredForeignKeys(schema ast.CIStr, tbInfo *model.TableInfo, schemaMetaVersion int64) {
+	for _, fk := range tbInfo.ForeignKeys {
+		if fk.Version < model.FKVersion1 {
+			continue
+		}
+		// found if already exists
+		referredFKInfos := isd.getTableReferredForeignKeys(fk.RefSchema.L, fk.RefTable.L, schemaMetaVersion)
+		found := false
+		for _, referredFK := range referredFKInfos {
+			if referredFK.ChildSchema.L == schema.L && referredFK.ChildTable.L == tbInfo.Name.L && referredFK.ChildFKName.L == fk.Name.L {
+				found = true
+				break
+			}
+		}
+		if !found {
+			continue
+		}
```

Comment:
Repeated code can be extracted into a function, noticed the same pattern in addReferredForeignKeys?

Reply:
[c45a3a4](https://github.com/pingcap/tidb/pull/59422/commits/c45a3a4904a74cdd0b4b110b3c7a47899909a4ff)